### PR TITLE
 Support vpi_put/vpi_get forcing of signals (#5933)

### PR DESCRIPTION
**Pull request by Christian-Hecken**
This commit includes PR https://github.com/verilator/verilator/pull/6688. It adds support for `vpiForceFlag` and `vpiReleaseFlag` to `vpi_put_value`. `vpi_get_value` is updated to return the value of either the signal itself, or the `__VforceVal` control signal, depending on whether the forcing is active.

The functionality is tested in the tests `t_vpi_force`, which tests that forcing and releasing works on a clocked register being forced with a VpiIntVal, as well as `t_vpi_forceable_bad`, which tests that attempting to force a signal without marking it forceable causes an error. The tests were run under Verilator (based on dc00bf248413c0c40c3f144764691e7e0ed38009 with the commit for isForceable applied), `Icarus Verilog version 13.0 (devel) (s20251012-20-gcc496c3cf)`, and `xrun 24.09-a071`.

The implementation is simply done using string concatenation of the signal name with the __VforceEn and __VforceVal suffixes. The signal vop that the vpi_put/get functions operate on is then pointed towards either the base signal or the __VforceVal signal (for vpi_put) or the __VforceRd signal (for vpi_get).

While it works and passes the newly implemented tests, this solution is quite brittle and in part re-implements existing functionality by recreating the `__VforceRd` signal, so once https://github.com/verilator/verilator/pull/6705 is completed, it should extend `VerilatedVar` to hold information about the force control signals, which is provided at Verilation time, rather than a runtime lookup.

Because `valuep` should get set to the value of the signal after forcing when `vpi_put_value` is called with `vpiReleaseFlag`, this
commit also adds information about `isContinuously` to the `VerilatedVarFlags`.

Lastly, since unpacked arrays cannot be forced (https://github.com/verilator/verilator/issues/4735), a Verilation time check for this was added and tested in `t_forceable_unpacked_bad`, which simplifies the error handling in `vpi_put_value` and `vpi_get_value`. The same was done for forceable strings, which is tested in `t_forceable_string_bad`.

Fixes https://github.com/verilator/verilator/issues/5933

### CODE REVIEW

File: src/V3AstNodes.cpp [hunk 1]
```diff
@@ -628,6 +628,7 @@ string AstVar::vlEnumDir() const {
     } else if (isSigUserRdPublic()) {
         out += "|VLVF_PUB_RD";
     }
+    if (isForceable()) { out += "|VLVF_FORCEABLE"; }
```
Comment (wsnyder) [inline src/V3AstNodes.cpp]:
```suggestion
    if (isForceable()) out += "|VLVF_FORCEABLE";
```

File: test_regress/t/t_vpi_forceable_bad.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import vltest_bootstrap
+
+test.scenarios('simulator')
+
+test.compile(make_top_shell=False,
+             make_main=False,
+             make_pli=True,
+             verilator_flags2=["--exe --vpi", test.pli_filename])
+
+test.execute(use_libvpi=True, fails=test.vlt,
```
Comment (wsnyder) [inline test_regress/t/t_vpi_forceable_bad.py]: Please add an .out file that looks for the appropriate message.  (Otherwise something else might break in future to make it still fail but wrongly.)

```suggestion
test.execute(use_libvpi=True, fails=test.vlt, expect_filename=test.golden_filename,
```

File: test_regress/t/t_vpi_forceable_bad.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
+#ifdef IS_VPI
+
```
Comment (wsnyder) [inline test_regress/t/t_vpi_forceable_bad.cpp]: Not sure why ifdef is needed.   Add standard copyright header.  (Same other files.)

File: test_regress/t/t_vpi_forceable_bad.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    for (s_vpi_systf_data& systf : vpi_systf_data) vpi_register_systf(&systf);
+}
+
+// icarus entry
+void (*vlog_startup_routines[])() = {vpi_compat_bootstrap, 0};
+
+#else
+int main(int argc, char** argv) {
+    const std::unique_ptr<VerilatedContext> contextp{new VerilatedContext};
+    constexpr uint64_t sim_time = 5;
```
Comment (wsnyder) [inline test_regress/t/t_vpi_forceable_bad.cpp]: Suggest pass `--main` to verilator so don't need this.  (Same other files.)

File: test_regress/t/t_vpi_force_minimal.cpp [hunk 5]
```diff
@@ -0,0 +1,232 @@
+#ifdef IS_VPI
+
```
Comment (wsnyder) [inline test_regress/t/t_vpi_force_minimal.cpp]: Why _minimal?  Seems fairly complete and will grow ;)

File: test_regress/t/t_vpi_force.v [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+`ifdef VERILATOR
+`systemc_header
+extern "C" int baselineValue();
+extern "C" int forceValues();
+extern "C" int releaseValues();
+extern "C" int checkValuesForced();
+extern "C" int checkValuesReleased();
+`verilog
+`endif
```
Comment (wsnyder) [inline test_regress/t/t_vpi_force.v:60]: Suggest use DPI imports, then should work with all simulators.

File: test_regress/t/t_vpi_force_minimal.v [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+  integer vpiStatus;
+
+  initial begin
+`ifndef VERILATOR
+`ifdef WAVES
+    $dumpfile(`STRINGIFY(`TEST_DUMPFILE));
+    $dumpvars();
+`endif
+`endif
```
Comment (wsnyder) [inline test_regress/t/t_vpi_force_minimal.v]: Verilator supports these, no ifdef needed.

File: include/verilated_vpi.cpp [hunk 8]
```diff
@@ -2593,6 +2593,45 @@ void vl_vpi_put_word(const VerilatedVpioVar* vop, QData word, size_t bitCount, s
     }
 }
 
+namespace {
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]: Put in VerilatedVpiImp instead

File: include/verilated_vpi.cpp [hunk 9]
```diff
@@ -2593,6 +2593,45 @@ void vl_vpi_put_word(const VerilatedVpioVar* vop, QData word, size_t bitCount, s
...
+
+// vpi_handle_by_name expects a non-const PLI_BYTE8* array, so this function creates a copy of the
+// const string name.
+vpiHandle vpiHandleFromString(const std::string& name) {
+    std::vector<char> nameCopy(name.begin(), name.end());
+    nameCopy.push_back('\0');
+    return vpi_handle_by_name(reinterpret_cast<PLI_BYTE8*>(nameCopy.data()), nullptr);
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]:
```suggestion
    return vpi_handle_by_name(reinterpret_cast<PLI_BYTE8*>(nameCopy.c_str()), nullptr);
```

File: include/verilated_vpi.cpp [hunk 10]
```diff
@@ -2703,6 +2751,71 @@ void vl_vpi_get_value(const VerilatedVpioVarBase* vop, p_vpi_value valuep) {
             return;
         }
     } else if (valuep->format == vpiIntVal) {
+        if (VL_UNLIKELY(vop->varp()->isForceable())) {
+            // TODO: __VforceRd already has the correct value, but that signal is not available
+            const auto forceControlSignals = getForceControlSignals(vop);
+            const vpiHandle& forceEnableSignalp = forceControlSignals.first;
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]: Please refactor into something that we can better scale to support new formats.  E.g. similar to how the existing cose uses vl_vpi_put_word and has only 10 lines or so for each format.

For that matter, maybe go right for supporting arbitrary widths?

File: test_regress/t/t_vpi_force.v (hunk 6 above)
Comment (Christian-Hecken) [reply test_regress/t/t_vpi_force.v:60]: Icarus doesn't support it. Added `USE_VPI_NOT_DPI` flag to be able to select either.

File: include/verilated_vpi.cpp (hunk 9 above)
Comment (Christian-Hecken) [reply include/verilated_vpi.cpp]: I used the `push_back('\0')` workaround to avoid const casts, since `c_str()` is `const`. So to implement your `c_str()` suggestion, I just ditched `vpiHandleFromString` entirely.

File: test_regress/t/t_vpi_force.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python3
+# DESCRIPTION: vpi force and release test
+# This test checks that forcing a signal using vpi_put_value with vpiForceFlag
+# sets it to the correct value, and then releasing it with vpiReleaseFlag
+# returns it to the initial state. This is a basic test that just checks the
+# correct behavior for a clocked register being forced with a VpiIntVal.
```
Comment (wsnyder) [inline test_regress/t/t_vpi_force.py]: Please move comments to .cpp.

File: test_regress/t/t_vpi_force.v [hunk 12]
```diff
@@ -0,0 +1,170 @@
+// ======================================================================
+// This file ONLY is placed under the Creative Commons Public Domain, for
+// any use, without warranty.
+// SPDX-License-Identifier: CC0-1.0
+// ======================================================================
+`define STRINGIFY(x) `"x`"
```
Comment (wsnyder) [inline test_regress/t/t_vpi_force.v:7]:
```suggestion

`define STRINGIFY(x) `"x`"
```

File: test_regress/t/t_vpi_force.v [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+endmodule
+
+module Test (
+    input clk
+);
+
+  logic clockedReg  /*verilator public_flat_rw*/  /*verilator forceable*/, assignedWire;
+`ifdef VERILATOR
+  assign assignedWire = $c32("baselineValue()");
```
Comment (wsnyder) [inline test_regress/t/t_vpi_force.v]:
```suggestion
`ifndef USE_VPI_NOT_DPI
  assign assignedWire = baselineValue();
```

File: test_regress/t/t_vpi_forceable_bad.cpp [hunk 14]
```diff
@@ -0,0 +1,57 @@
+// ======================================================================
+// This file ONLY is placed under the Creative Commons Public Domain, for
+// any use, without warranty.
+// SPDX-License-Identifier: CC0-1.0
+// ======================================================================
+#include "verilated.h"
```
Comment (wsnyder) [inline test_regress/t/t_vpi_forceable_bad.cpp:14]:
```suggestion
// ======================================================================

#include "verilated.h"
```

File: test_regress/t/t_vpi_forceable_bad.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python3
+# DESCRIPTION: Test failure of trying to force a non-forceable signal
+# This test checks that attempting to force a signal that is not marked as
+# forceable causes an error under Verilator, and does not cause an error in
+# other simulators that do not need this metacomment to be able to force
+# signals.
+#
```
Comment (wsnyder) [inline test_regress/t/t_vpi_forceable_bad.py]: Likewise, put in .cpp or .v.

File: test_regress/t/t_vpi_forceable_bad.v [hunk 16]
```diff
@@ -0,0 +1,50 @@
+// ======================================================================
+// This file ONLY is placed under the Creative Commons Public Domain, for
+// any use, without warranty.
+// SPDX-License-Identifier: CC0-1.0
+// ======================================================================
+module t ();
```
Comment (wsnyder) [inline test_regress/t/t_vpi_forceable_bad.v]:
```suggestion
// ======================================================================

module t;
```

File: test_regress/t/t_vpi_forceable_bad.v [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+`endif
+
+  wire nonForceableSignal  /*verilator public_flat_rw*/ = 1'b0;
+  integer vpiStatus = 1;
+
+  initial begin
+
+`ifdef VERILATOR
+`ifdef USE_VPI_NOT_DPI
+    vpiStatus = $c32("forceValue()");
```
Comment (wsnyder) [inline test_regress/t/t_vpi_forceable_bad.v:27]:
```suggestion
    vpiStatus = forceValue();
```

File: include/verilated_vpi.cpp [hunk 18]
```diff
@@ -1075,6 +1076,23 @@ class VerilatedVpiImp final {
         }
         s().m_inertialPuts.clear();
     }
+    static std::pair<vpiHandle, vpiHandle> getForceControlSignals(const VerilatedVpioVarBase* vop);
+    static void setAllBitsToValue(const VerilatedVpioVar* vop, uint8_t bit_value) {
+        assert(bit_value == 0 || bit_value == 1);
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]:
```suggestion
    static void setAllBitsToValue(const VerilatedVpioVar* vop, uint8_t bitValue) {
        assert(bitValue == 0 || bitValue == 1);
```

File: include/verilated_vpi.cpp [hunk 19]
```diff
@@ -1075,6 +1076,23 @@ class VerilatedVpiImp final {
...
+        for (std::size_t i{0}; i < numChunks; ++i) {
+            vl_vpi_put_word(vop, word, wordSize, i * wordSize);
+        }
+        // addOffset == varBits would trigger assertion in vl_vpi_var_access_info even if
+        // bitCount == 0, so first check if there is a remainder
+        if (varBits % wordSize != 0)
+            vl_vpi_put_word(vop, word, varBits % wordSize, numChunks * wordSize);
+    }
+    static void setAllBits(const VerilatedVpioVar* vop) { setAllBitsToValue(vop, 1); }
+    static void clearAllBits(const VerilatedVpioVar* vop) { setAllBitsToValue(vop, 0); }
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]: Suggest remove these and inline where used, since just used once.

File: include/verilated_vpi.cpp [hunk 20]
```diff
@@ -1229,6 +1247,36 @@ VerilatedVpiError* VerilatedVpiImp::error_info() VL_MT_UNSAFE_ONE {
...
+    vpiHandle forceValueSignalp
+        = vpi_handle_by_name(const_cast<PLI_BYTE8*>(forceValueSignalName.c_str()), nullptr);
+    if (VL_UNLIKELY(!VerilatedVpioVar::castp(forceEnableSignalp))) {
+        VL_VPI_ERROR_(__FILE__, __LINE__,
+                      "%s: vpi force or release requested for '%s', but vpiHandle '%p' of control "
+                      "signal '%s' could not be cast to VerilatedVpioVar*. Ensure signal is "
+                      "marked as forceable",
+                      __func__, signalName.c_str(), forceEnableSignalp,
+                      forceEnableSignalName.c_str());
+        forceEnableSignalp = nullptr;
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]:
```suggestion
```

Is already null.  Same below.

File: include/verilated_vpi.cpp [hunk 21]
```diff
@@ -1229,6 +1247,36 @@ VerilatedVpiError* VerilatedVpiImp::error_info() VL_MT_UNSAFE_ONE {
...
+std::pair<vpiHandle, vpiHandle>
+VerilatedVpiImp::getForceControlSignals(const VerilatedVpioVarBase* const vop) {
+    const std::string signalName = vop->fullname();
+    const std::string forceEnableSignalName = signalName + "__VforceEn";
+    const std::string forceValueSignalName = signalName + "__VforceVal";
+
+    vpiHandle forceEnableSignalp
+        = vpi_handle_by_name(const_cast<PLI_BYTE8*>(forceEnableSignalName.c_str()), nullptr);
+    vpiHandle forceValueSignalp
+        = vpi_handle_by_name(const_cast<PLI_BYTE8*>(forceValueSignalName.c_str()), nullptr);
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]:
```suggestion
    vpiHandle const forceEnableSignalp
        = vpi_handle_by_name(const_cast<PLI_BYTE8*>(forceEnableSignalName.c_str()), nullptr);
    vpiHandle const forceValueSignalp
        = vpi_handle_by_name(const_cast<PLI_BYTE8*>(forceValueSignalName.c_str()), nullptr);
```

File: include/verilated_vpi.cpp [hunk 22]
```diff
@@ -2604,6 +2652,42 @@ void vl_vpi_get_value(const VerilatedVpioVarBase* vop, p_vpi_value valuep) {
 
     const int varBits = vop->bitSize();
 
+    // TODO: __VforceRd already has the correct value, but that signal is not public and thus not
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]:
```suggestion
    // __VforceRd already has the correct value, but that signal is not public and thus not
```

Presume this now works...

File: include/verilated_vpi.cpp [hunk 23]
```diff
@@ -2749,55 +2831,162 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
...
+                if (VL_UNLIKELY(errorOccurred && forceValueGetError.level < vpiError)) {
+                    vpi_printf(forceValueGetError.message);
+                    VL_VPI_ERROR_RESET_();
+                }
+            }
+
+            return object;  // TODO: According to the SystemVerilog specification,
+                            // vpi_put_value should return a handle to the scheduled event
+                            // if the vpiReturnEvent flag is selected, NULL otherwise. Is
+                            // this even possible with Verilator?
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]:
```suggestion
```suggestion
            // TODO: According to the SystemVerilog specification,
            // vpi_put_value should return a handle to the scheduled event
            // if the vpiReturnEvent flag is selected, NULL otherwise.
            return object;  
```

I'm sure we can figure out a way, but we can leave for later.

**Comment by wsnyder** [comment]
Enabling coverage - please check that all lines you added are covered, it's ok if there's uncovered branches you think are not critical.

**Comment by wsnyder** [comment]
@Christian-Hecken how's this looking, is it ready for another review?  If so can you click "resolved" on all of the feedback items that seem appropriate.

Note to myself, this also seems to be needed for uvm_tests tests/70regs/30bkdr/01hdl.

These are forcing the below. Thus would prefer this patch handle at least all signals <= 64 bits.

  ```
   wire [7:0] w;
   wire [7:0] w_nodrv;
   reg [7:0] q = 'h0F;
   reg [7:0] d = 'hF0;
```

File: test_regress/t/t_vpi_forceable_bad.v (hunk 17 above)
Comment (Christian-Hecken) [reply test_regress/t/t_vpi_forceable_bad.v:27]: Not sure what you mean - I was under the impression that vpi calls are supposed to be made with the `$c32` syntax, and it's just DPI that doesn't need it.

File: include/verilated_vpi.cpp (hunk 22 above)
Comment (Christian-Hecken) [reply include/verilated_vpi.cpp]: Well, no. That's what #6705 is for. I edited the comment to make this clearer.

**Comment by Christian-Hecken** [comment]
Apologies for the delay.
While improving test coverage, several issues showed up requiring some architectural changes and significantly expanding the testbench. The upside is that both `vpi_put_value` and `vpi_get_value` are now fully featured with regard to forcing, including continuously assigned signals and all data types. Future changes in #6705 should then only affect performance and code quality, but add no more features.
Support for continuously and non-continuously assigned signals required a simple addition to the VerilatedVar's `m_vlflags`, and to simplify error handling in `vpi_put_value` and `vpi_get_value`, forceable unpacked arrays are now rejected at Verilation time, since they are unsupported (#4735).

The tests show that Verilator's behavior matches Icarus and Xcelium, with the exception of a continuously assigned `real`, where `xrun` returns the _forced_ value as the output `s_vpi_val` when releasing with `vpiReleaseFlag` through `vpi_put_value`, rather than the expected _release_ value, like with all other continuously assigned variables. For now, I ensured the tests still pass by inserting a check for this specific variable if the test is run with `xrun`, but I am unsure which behavior is correct here.
In order to get to 100% line coverage, I had to cheat and exclude some edge cases from coverage for which I could not figure out a way to test them. `VL_UNCOVERABLE` also did not appear to work properly, so I used `LCOV_EXCL` instead. Please let me know if this is still sufficient. I am also far from happy with the branch coverage, but I suspect pushing that to 100% as well is going to be a substantial amount of effort, so I'd like to check in with you first which target you propose.

Currently, `vpi_get_value` does not return the correct value immediately after a `vpi_put_value`, because when manually re-creating the effect of the `__VforceRd` signal in `vpi_get_value`, it sees any changes of the `__VforceEn` and `__VforceVal` signals _instantly_, whereas the actual `__VforceRd` signal only gets updated in the `eval_act` step of `eval_phase__act`, which occurs _after_ the `timing_resume` function gets called, which contains the `vpi_put_value` and `vpi_get_value` in `t_vpi_force`. This should get fixed in #6705, and I am not sure if this is so critical as to warrant cancelling this PR entirely and waiting for #6705 to be finished.

Lastly, it seems that Verilator supports forcing strings, but as far as I can tell, that is not allowed by the SystemVerilog standard, and also not supported by Icarus and Xcelium. This is why I added a check in `vpi_get_value` for attempting to retrieve the value of a forceable string, but a better solution would be to already raise an error at Verilation time. Before adding this though, I wanted to check if this is an intentional feature in Verilator and should remain.

File: test_regress/t/t_vpi_forceable_bad.v (hunk 17 above)
Comment (wsnyder) [reply test_regress/t/t_vpi_forceable_bad.v:27]: Correct, just DPI doesn't need it, but why not always use DPI so it's the same code for Verilator or others?

**Comment by wsnyder** [comment]
Thanks, I agree with all your comments.  Note that strictly IEEE allows only forcing of net types (6.5), and net types must be 4-state compatible (6.7.1), thus strings are not legal to force per IEEE.  Verilator is much more permissive as you noted.

File: test_regress/t/t_vpi_force.py [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+
+import vltest_bootstrap
+
+test.scenarios('simulator')
+
+test.compile(make_top_shell=False,
+             make_main=False,
+             make_pli=True,
+             verilator_flags2=["--main --exe --vpi --timing", test.pli_filename],
```
Comment (wsnyder) [inline test_regress/t/t_vpi_force.py]: `--binary` == `--main --exe --timing`.  Similar on others.

```suggestion
             verilator_flags2=["--binary --vpi", test.pli_filename],
```

File: test_regress/t/t_vpi_forceable_unpacked_bad.py [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+# can redistribute it and/or modify it under the terms of either the GNU
+# Lesser General Public License Version 3 or the Perl Artistic License
+# Version 2.0.
+# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+
+import vltest_bootstrap
+
+test.scenarios('simulator')
+
+test.compile(fails=test.vlt_all, expect_filename=test.golden_filename)
```
Comment (wsnyder) [inline test_regress/t/t_vpi_forceable_unpacked_bad.py:14]:
```suggestion
test.scenarios('vlt_st')

test.lint(fails=test.vlt_all, expect_filename=test.golden_filename)
```

File: src/V3Force.cpp [hunk 26]
```diff
@@ -323,6 +323,12 @@ class ForceConvertVisitor final : public VNVisitor {
     void visit(AstVarScope* nodep) override {
         // If this signal is marked externally forceable, create the public force signals
         if (nodep->varp()->isForceable()) {
+            if (VN_IS(nodep->varp()->dtypeSkipRefp(), UnpackArrayDType)) {
+                nodep->varp()->v3error(
+                    "Forcing unpacked arrays is not supported (#4735): " << nodep->varp()->name());
```
Comment (wsnyder) [inline src/V3Force.cpp]:
```suggestion
                nodep->varp()->v3warn(UNSUPPORTED,
                    "Unsupported: Forcing unpacked arrays: " << nodep->varp()->name());  // (#4735)
```

File: include/verilated_vpi.cpp [hunk 27]
```diff
@@ -2651,62 +2807,73 @@ void vl_vpi_get_value(const VerilatedVpioVarBase* vop, p_vpi_value valuep) {
...
             t_outDynamicStr[chars - i - 1] = "0123456789abcdef"[static_cast<int>(val)];
         }
         valuep->value.str = const_cast<PLI_BYTE8*>(t_outDynamicStr.c_str());
         return;
     } else if (valuep->format == vpiStringVal) {
         if (varp->vltype() == VLVT_STRING) {
+            if (VL_UNLIKELY(varp->isForceable())) {
+                VL_VPI_ERROR_(__FILE__, __LINE__,
+                              "attempting to retrieve value of forceable signal %s with data type "
+                              "VLVT_STRING, but strings cannot be forced.",
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]:
```suggestion
                              "Attempting to retrieve value of forceable signal '%s' with data type "
                              "VLVT_STRING, but strings cannot be forced.",
```

File: include/verilated_vpi.cpp [hunk 28]
```diff
@@ -2749,55 +2916,158 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
...
+                          baseSignalVop->fullname(), baseSignalVop->scopep()->defname());
             return nullptr;
         }
-        if (!vl_check_format(vop->varp(), valuep, vop->fullname(), false)) return nullptr;
+
+        // NOLINTNEXTLINE(readability-simplify-boolean-expr);
+        if (VL_UNLIKELY((forceFlag == vpiForceFlag || forceFlag == vpiReleaseFlag)
+                        && !baseSignalVop->varp()->isForceable())) {
+            VL_VPI_ERROR_(__FILE__, __LINE__,
+                          "vpi_put_value was used with %s on non-forceable signal %s : %s",
```
Comment (wsnyder) [inline include/verilated_vpi.cpp]:
```suggestion
                          "vpi_put_value was used with %s on non-forceable signal '%s': '%s'",
```

File: test_regress/t/t_vpi_forceable_unpacked_bad.py (hunk 25 above)
Comment (Christian-Hecken) [reply test_regress/t/t_vpi_forceable_unpacked_bad.py:14]: `vlt_st` does not appear to be a valid scenario `driver.py` checks for. I used `vlt` instead.

**Comment by Christian-Hecken** [comment]
I resolved your review comments and added the Verilation time check for forceable strings, since you said that strings are not legal to force per IEEE. If I misunderstood you and Verilator being more permissive is a feature, not a bug, then I can remove that last commit again.

File: test_regress/t/t_vpi_forceable_bad.v (hunk 17 above)
Comment (Christian-Hecken) [reply test_regress/t/t_vpi_forceable_bad.v:27]: Because `driver.py` does not currently enable `xrun` to use DPI, and only using VPI for `xrun` and DPI for all other simulators would seem arbitrary and confusing for future readers, so I added support for both methods for all simulators.

**Comment by wsnyder** [comment]
Is this good to merge?  I'll run the rtlmeter just in case it trips on something which is unlikely.

**Comment by Christian-Hecken** [comment]
Yes, I think it should be ready.

**Comment by wsnyder** [comment]
Realized we're missing a piece, can you please update these files to use the force?

```
test_regress/t/t_uvm_dpi.v  (Add real test where see uvm_hdl_force/uvm_hdl_release)
test_regress/t/uvm/v2017_1_0/dpi/uvm_hdl_verilator.c
test_regress/t/uvm/v2020_3_1/dpi/uvm_hdl_verilator.c   (will be same edit as for above)
```

**Comment by wsnyder** [comment]
@Christian-Hecken do you think you'll be able to polish this off?  Would be good to get in, as some UVM tests would pass with it.

**Comment by Christian-Hecken** [comment]
Unfortunately, changing the UVM tests isn't as simple as it seems. The current implementation of VPI forcing requires an eval cycle between a write and a read of a forceable signal. There's no such eval cycle in the UVM test, hence we cannot simply change the signals to be forceable and have the tests still pass.

I won't get around to improve the code to not need the eval cycle in the next couple of weeks, unfortunately, as I'm focusing on a report for university first. I'm happy to take a look afterwards. I leave it up to you if you'd like to keep this PR open until we have the extended functionality in place, or merge this one as-is and I'll file a follow up afterwards.

**Comment by wsnyder** [comment]
Merged manually, thanks.  We'll deal with the UVM stuff later.
